helix-core/src/syntax.rs

```diff
@@ -1809,24 +1826,62 @@ impl<'a> Iterator for HighlightIter<'a> {
     }
 }
 
+#[derive(Debug, Clone)]
+pub enum InjectionCapture<'a> {
+    Name(Cow<'a, str>),
+    File(Cow<'a, Path>),
+    Shebang(Rope),
+}
+
+static SHEBANG: &str = r"#!\s*(?:\S*[/\\](?:env\s+(?:\-\S+\s+)*)?)?([^\s\.\d]+)";
+
 fn injection_for_match<'a>(
     config: &HighlightConfiguration,
     query: &'a Query,
     query_match: &QueryMatch<'a, 'a>,
     source: RopeSlice<'a>,
-) -> (Option<Cow<'a, str>>, Option<Node<'a>>, IncludedChildren) {
+) -> (
+    Option<InjectionCapture<'a>>,
+    Option<Node<'a>>,
+    IncludedChildren,
+) {
     let content_capture_index = config.injection_content_capture_index;
     let language_capture_index = config.injection_language_capture_index;
+    let language_filename_index = config.injection_filename_capture_index;
+    let language_shebang_index = config.injection_shebang_capture_index;
 
-    let mut language_name = None;
+    let mut language_capture = None;
     let mut content_node = None;
     for capture in query_match.captures {
         let index = Some(capture.index);
         if index == language_capture_index {
             let name = byte_range_to_str(capture.node.byte_range(), source);
-            language_name = Some(name);
+            language_capture = Some(InjectionCapture::Name(name));
         } else if index == content_capture_index {
             content_node = Some(capture.node);
+        } else if index == language_filename_index {
+            let name = byte_range_to_str(capture.node.byte_range(), source);
+            let path = Path::new(name.as_ref()).to_path_buf();
+
+            language_capture = Some(InjectionCapture::File(path.into()));
+        } else if index == language_shebang_index {
+            // assume that the content with the shebang is also the content to highlight
+            content_node = Some(capture.node);
```

Comment:
A single node can have multiple captures

```scm
(keyword) @injection.shebang @injection.content
```

so assuming the content node is not necessary here

Reply:
Indeed, this is how I first had it, but I just figured, when would we ever want to capture a shebang but not highlight the same node? I don't mind being more explicit though.